src/lib/dom-api-mutation.html

```diff
@@ -0,0 +1,280 @@
+The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+Code distributed by Google as part of the polymer project is also
+subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+-->
+<link rel="import" href="settings.html">
+<script>
+(function() {
+  'use strict';
+
+    var DomApi = Polymer.DomApi.ctor;
+    var Settings = Polymer.Settings;
+    var hasDomApi = Polymer.DomApi.hasDomApi;
+
+    DomApi.Mutation = function(domApi) {
+      this.domApi = domApi;
+      this.node = this.domApi.node;
+      this._listeners = [];
+      this._addedNodes = [];
+      this._removedNodes = [];
+    };
+
+    DomApi.Mutation.prototype = {
+
+      addListener: function(callback, options) {
+        options = options || {};
+        this._ensureSetup(options);
+        var listener = {fn: callback, options: options};
+        this._notifyInitial(listener);
+        return this._listeners.push(listener);
+      },
+
+      removeListener: function(handle) {
+        this._listeners.splice(handle - 1, 1);
+        if (!this._hasListeners()) {
+          this._ensureCleanup();
+        }
+      },
+
+      _ensureSetup: function(options) {
+        if (!this._isSetup) {
+          this._isSetup = true;
+          this._observeContentElements(this.domApi.childNodes);
+        }
+      },
+
+      _ensureCleanup: function() {}, // abstract
+
+      _hasListeners: function() {
+        return Boolean(this._listeners.length);
+      },
+
+      addNode: function(node) {
+        if (this._hasListeners()) {
+          this._addedNodes.push(node);
+          this._scheduleNotify();
+        }
+      },
+
+      removeNode: function(node) {
+        if (this._hasListeners()) {
+          this._removedNodes.push(node);
+          this._scheduleNotify();
+        }
+      },
+
+      addAllNodes: function(force) {
+        if (this._hasListeners() || force) {
+          var c$ = this.domApi.childNodes;
+          if (c$.length) {
+            for (var i=0, c; (i < c$.length) && (c=c$[i]); i++) {
+              this._addedNodes.push(c);
+            }
+            this._scheduleNotify();
+          }
+        }
+      },
+
+      _scheduleNotify: function() {
+        this._debouncer = Polymer.Debounce(this._debouncer, 
+          this._notify);
+        this._debouncer.context = this;
+        Polymer.dom.addDebouncer(this._debouncer);
+      },
+
+      _notify: function(mxns) {
+        var info = mxns || {
+          target: this.node,
+          addedNodes: this._addedNodes,
+          removedNodes: this._removedNodes
+        }
+        this._updateContentElements(info);
+        this._callListeners(info);
+        this._addedNodes = [];
+        this._removedNodes = [];
+      },
+
+      _notifyInitial: function(listener) {
+        var info = {

```

Comment:
Only go if there are `childNodes`?  Otherwise will get an empty notification followed by the one from shady's call to `addAllNodes` if observed before distribution?